man/caffe.1
.\"Text automatically generated by txt2man
.TH CAFFE 1 "10 August 2016" "" ""
.SH NAME
\fBcaffe \fP- command line brew for Caffe
.SH SYNOPSIS
.nf
.fam C
 \fBcaffe\fP <COMMAND> <FLAGS>
.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
Caffe is a deep learning framework made with expression, speed, 
and modularity in mind. It is developed by the Berkeley Vision
and Learning Center (BVLC) and community contributors.
.SH COMMANDS
.TP
.B
train
train or finetune a model
.TP
.B
test
score a model
.TP
.B
device_query
show GPU diagnostic information
.TP
.B
time
benchmark model execution time
.SH FREQUENTLY USED FLAGS
.TP
.B
\fB-gpu\fP
(Optional; run in GPU mode on given device IDs separated by ','.
Use '\fB-gpu\fP all' to run on all available GPUs. The effective
training batch size is multiplied by the number of devices.)
type: string default: ""
.TP
.B
\fB-iterations\fP
(The number of iterations to run.) type: int32 default: 50
.TP
.B
\fB-level\fP
(Optional; network level.) type: int32 default: 0
.TP
.B
\fB-model\fP
(The model definition protocol buffer text file..) type: string
default: ""
.TP
.B
\fB-phase\fP
(Optional; network phase (TRAIN or TEST). Only used for 'time'.)
type: string default: ""
.TP
.B
\fB-sighup_effect\fP
(Optional; action to take when a SIGHUP signal is received:
snapshot, stop or none.) type: string default: "snapshot"
.TP
.B
\fB-sigint_effect\fP
(Optional; action to take when a SIGINT signal is received:
snapshot, stop or none.) type: string default: "stop"
.TP
.B
\fB-snapshot\fP
(Optional; the snapshot solver state to resume training.)
type: string default: ""
.TP
.B
\fB-solver\fP
(The solver definition protocol buffer text file.) type: string
default: ""
.TP
.B
\fB-stage\fP
(Optional; network stages (not to be confused with phase), separated
by ','.) type: string default: ""
.TP
.B
\fB-weights\fP
(Optional; the pretrained weights to initialize finetuning,
separated by ','. Cannot be set simultaneously with snapshot.)
type: string default: ""
.TP
.B
\fB-help\fP
Show complete help messages.
.SH OTHER CAFFE UTILITIES
Apart from the "\fBcaffe\fP" command line utility, there are also some utilities
available, run them with "\fB-h\fP" or "\fB--help\fP" argument to see corresponding help.
.IP \(bu 3
convert_imageset
.IP \(bu 3
convert_cifar_data
.IP \(bu 3
compute_image_mean
.IP \(bu 3
convert_mnist_siamese_data
.IP \(bu 3
upgrade_net_proto_binary
.IP \(bu 3
extract_features
.IP \(bu 3
upgrade_solver_proto_text
.IP \(bu 3
classification
.IP \(bu 3
upgrade_net_proto_text
.IP \(bu 3
convert_mnist_data
.SH EXAMPLES
Train a new Network
.PP
.nf
.fam C
    $ caffe train -solver solver.prototxt

.fam T
.fi
Resume training a network from a snapshot
.PP
.nf
.fam C
    $ caffe train -solver solver.prototxt -snapshot bvlc_alexnet.solverstate

.fam T
.fi
Fine-tune a network
.PP
.nf
.fam C
    $ caffe train -solver solver.prototxt -weights pre_trained.caffemodel

.fam T
.fi
Test (evaluate) a trained model for 100 iterations, on GPU 0
.PP
.nf
.fam C
    $ caffe test -model train_val.prototxt -weights bvlc_alexnet.caffemodel -gpu 0 -iterations 100

.fam T
.fi
Run a benchmark against AlexNet on GPU 0
.PP
.nf
.fam C
    $ caffe time -model deploy.prototxt -gpu 0

.fam T
.fi
Check CUDA device availability of GPU 0
.PP
.nf
.fam C
    $ caffe device_query -gpu 0

.fam T
.fi
.SH HOMEPAGE
http://caffe.berkeleyvision.org
.SH BUGS
https://github.com/BVLC/\fBcaffe\fP/issues
.SH AUTHOR
This manpage is written by Zhou Mo <cdluminate@gmail.com> with the help of txt2man for Debian
according to program's help message.


docs/release_notes.md

---
title: Release Notes
---
#Table Of Contents
- [Introduction](#Introduction)
- [Installation](#Installation)
 - [Prerequisites](#Prerequisites)
 - [Building for Intel® Architecture](#Building)
 - [Building for GPU](#Building)
 - [Compilation](#Compilation)
- [Configurations](#Configurations)
 - [Hardware](#hardware)
 - [Software](#Software)
- [Known issues and limitations](#Known)
- [Instructions](#Instructions)
 - [How to measure performance](#performance)
 - [How to train singlenode](#singlenode)
 - [How to train multinode](#multinode)
 - [How to contribute](#contribute)
- [License](#License)


# Introduction

This fork is dedicated to improving Caffe performance when running on CPU, in particular Intel® Xeon processors (HSW, BDW+)

# Installation

Prior to installing, have a glance through this guide and take note of the details for your platform.
We build and test Caffe on CentOS (7.0, 7.1, 7.2). 
The official Makefile and `Makefile.config` build are complemented by an automatic CMake build from the community.

When updating Caffe, it's best to `make clean` before re-compiling.

## Prerequisites

Before building Caffe make sure that the following dependencies are available on target system:

* [Boost](http://www.boost.org/) >= 1.55
* `protobuf`, `glog`, `gflags`, `hdf5`

For additional capabilities and acceleration the following dependencies might be necessary:

* [OpenCV](http://opencv.org/) >= 2.4 including 3.0
* IO libraries: `lmdb`, `leveldb` (note: leveldb requires `snappy`)
* For GPU mode
    * [CUDA](https://developer.nvidia.com/cuda-zone)
    * cuDNN

* For Pycaffe
    * `Python 2.7` or `Python 3.3+`
    * `numpy (>= 1.7)`
    * boost-provided `boost.python`

* For Matcaffe 
  * MATLAB with the `mex` compiler.

### Python and/or MATLAB Caffe (optional)

#### Python

The main requirements are `numpy` and `boost.python` (provided by boost). `pandas` is useful too and needed for some examples.

You can install the dependencies with

    for req in $(cat requirements.txt); do pip install $req; done

but we suggest first installing the [Anaconda](https://store.continuum.io/cshop/anaconda/) Python distribution, which provides most of the necessary packages, as well as the `hdf5` library dependency.

To import the `caffe` Python module after completing the installation, add the module directory to your `$PYTHONPATH` by `export PYTHONPATH=/path/to/caffe/python:$PYTHONPATH` or the like. You should not import the module in the `caffe/python/caffe` directory!

*Caffe's Python interface works with Python 2.7. Python 3.3+ should work out of the box without protobuf support. For protobuf support please install protobuf 3.0 alpha (https://developers.google.com/protocol-buffers/). Earlier Pythons are your own adventure.*

#### MATLAB

Install MATLAB, and make sure that its `mex` is in your `$PATH`.

*Caffe's MATLAB interface works with versions 2015a, 2014a/b, 2013a/b, and 2012b.*



##Building for Intel® Architecture

This version of Caffe is optimized for Intel® Xeon processors and Intel® Xeon Phi™ processors. To achieve the best performance results on Intel Architecture we recommend building Caffe with [Intel® MKL](http://software.intel.com/en-us/intel-mkl) and enabling OpenMP support. 
This Caffe version is seflcontained. This means that newest version of Intel MKL will be downloaded and installed during compilation of IntelCaffe.

* Set `BLAS := mkl` in `Makefile.config`
* If you don't need GPU optimizations `CPU_ONLY := 1` flag in `Makefile.config` to configure and build Caffe without CUDA.

[Intel MKL 2017 Beta Update 1](https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/623305) introduces optimized Deep Neural Network (DNN) performance primitives that allow to accelerate the most popular image recognition topologies. Caffe can take advantage of these primitives and get significantly better performance results compared to the previous versions of Intel MKL. There are two ways to take advantage of the new primitives: 


tools/caffe.cpp
#ifdef WITH_PYTHON_LAYER
#include "boost/python.hpp"
namespace bp = boost::python;
#endif

#include <gflags/gflags.h>
#include <glog/logging.h>

#include <cstring>
#include <map>
#include <string>
#include <vector>

#include "boost/algorithm/string.hpp"
#include "boost/make_shared.hpp"
#include "caffe/caffe.hpp"
#include "caffe/internode/mpiutil.hpp"
#include "caffe/multinode/multinode.hpp"
#include "caffe/util/signal_handler.h"

using caffe::Blob;
using caffe::Caffe;
using caffe::Net;
using caffe::Layer;
using caffe::Solver;
using caffe::shared_ptr;
using caffe::string;
using caffe::Timer;
using caffe::vector;
using std::ostringstream;

DEFINE_string(gpu, "",
    "Optional; run in GPU mode on given device IDs separated by ','."
    "Use '-gpu all' to run on all available GPUs. The effective training "
    "batch size is multiplied by the number of devices.");
DEFINE_string(solver, "",
    "The solver definition protocol buffer text file.");
DEFINE_string(model, "",
    "The model definition protocol buffer text file.");
DEFINE_string(phase, "",
    "Optional; network phase (TRAIN or TEST). Only used for 'time'.");
DEFINE_int32(level, 0,
    "Optional; network level.");
DEFINE_string(stage, "",
    "Optional; network stages (not to be confused with phase), "
    "separated by ','.");
DEFINE_string(snapshot, "",
    "Optional; the snapshot solver state to resume training.");
DEFINE_string(weights, "",
    "Optional; the pretrained weights to initialize finetuning, "
    "separated by ','. Cannot be set simultaneously with snapshot.");
DEFINE_int32(iterations, 50,
    "The number of iterations to run.");
DEFINE_string(sigint_effect, "stop",
             "Optional; action to take when a SIGINT signal is received: "
              "snapshot, stop or none.");
DEFINE_string(sighup_effect, "snapshot",
             "Optional; action to take when a SIGHUP signal is received: "
             "snapshot, stop or none.");
DEFINE_string(param_server, "",
    "Optional; triggers multinode mode, usage: --param_server=mpi");
DEFINE_string(listen_address, "",
    "Optional; multinode mode, bind address for data server");
DEFINE_int32(comm_threads, 1,
    "Optional; multinode mode,"
    " The number of threads used by communication code.");

// A simple registry for caffe commands.
typedef int (*BrewFunction)();
typedef std::map<caffe::string, BrewFunction> BrewMap;
BrewMap g_brew_map;

#define RegisterBrewFunction(func) \
namespace { \
class __Registerer_##func { \
 public: /* NOLINT */ \
  __Registerer_##func() { \
    g_brew_map[#func] = &func; \
  } \
}; \
__Registerer_##func g_registerer_##func; \
}

static BrewFunction GetBrewFunction(const caffe::string& name) {
  if (g_brew_map.count(name)) {
    return g_brew_map[name];
  } else {
    LOG(ERROR) << "Available caffe actions:";
    for (BrewMap::iterator it = g_brew_map.begin();
         it != g_brew_map.end(); ++it) {
      LOG(ERROR) << "\t" << it->first;
    }
    LOG(FATAL) << "Unknown action: " << name;
    return NULL;  // not reachable, just to suppress old compiler warnings.
  }
}

// Parse GPU ids or use all available devices
static void get_gpus(vector<int>* gpus) {
  if (FLAGS_gpu == "all") {
    int count = 0;
#ifndef CPU_ONLY
    CUDA_CHECK(cudaGetDeviceCount(&count));
#else
    NO_GPU;
#endif
    for (int i = 0; i < count; ++i) {
      gpus->push_back(i);
    }
  } else if (FLAGS_gpu.size()) {
    vector<string> strings;
    boost::split(strings, FLAGS_gpu, boost::is_any_of(","));
    for (int i = 0; i < strings.size(); ++i) {
      gpus->push_back(boost::lexical_cast<int>(strings[i]));
    }
  } else {
    CHECK_EQ(gpus->size(), 0);
  }
}

// Parse phase from flags
caffe::Phase get_phase_from_flags(caffe::Phase default_value) {
  if (FLAGS_phase == "")
    return default_value;
  if (FLAGS_phase == "TRAIN")
    return caffe::TRAIN;
  if (FLAGS_phase == "TEST")
    return caffe::TEST;
  LOG(FATAL) << "phase must be \"TRAIN\" or \"TEST\"";
  return caffe::TRAIN;  // Avoid warning
}

// Parse stages from flags
vector<string> get_stages_from_flags() {
  vector<string> stages;
  boost::split(stages, FLAGS_stage, boost::is_any_of(","));
  return stages;
}

// caffe commands to call by
//     caffe <command> <args>
//
// To add a command, define a function "int command()" and register it with
// RegisterBrewFunction(action);

// Device Query: show diagnostic information for a GPU device.
int device_query() {
  LOG(INFO) << "Querying GPUs " << FLAGS_gpu;
  vector<int> gpus;
  get_gpus(&gpus);
  for (int i = 0; i < gpus.size(); ++i) {
    caffe::Caffe::SetDevice(gpus[i]);
    caffe::Caffe::DeviceQuery();
  }
  return 0;
}
RegisterBrewFunction(device_query);

// Load the weights from the specified caffemodel(s) into the train and
// test nets.
void CopyLayers(caffe::Solver<float>* solver, const std::string& model_list) {
  std::vector<std::string> model_names;
  boost::split(model_names, model_list, boost::is_any_of(",") );
  for (int i = 0; i < model_names.size(); ++i) {
    LOG(INFO) << "Finetuning from " << model_names[i];
    solver->net()->CopyTrainedLayersFrom(model_names[i]);
    for (int j = 0; j < solver->test_nets().size(); ++j) {
      solver->test_nets()[j]->CopyTrainedLayersFrom(model_names[i]);
    }
  }
}

// Translate the signal effect the user specified on the command-line to the
// corresponding enumeration.
caffe::SolverAction::Enum GetRequestedAction(
    const std::string& flag_value) {
  if (flag_value == "stop") {
    return caffe::SolverAction::STOP;
  }
  if (flag_value == "snapshot") {
    return caffe::SolverAction::SNAPSHOT;
  }
  if (flag_value == "none") {
    return caffe::SolverAction::NONE;
  }
  LOG(FATAL) << "Invalid signal effect \""<< flag_value << "\" was specified";
}

// Train / Finetune a model.
int train() {
  CHECK_GT(FLAGS_solver.size(), 0) << "Need a solver definition to train.";
  CHECK(!FLAGS_snapshot.size() || !FLAGS_weights.size())
      << "Give a snapshot to resume training or weights to finetune "
      "but not both.";
  vector<string> stages = get_stages_from_flags();

  caffe::SolverParameter solver_param;
  caffe::ReadSolverParamsFromTextFileOrDie(FLAGS_solver, &solver_param);

  solver_param.mutable_train_state()->set_level(FLAGS_level);
  for (int i = 0; i < stages.size(); i++) {
    solver_param.mutable_train_state()->add_stage(stages[i]);
  }

  // If the gpus flag is not provided, allow the mode and device to be set
  // in the solver prototxt.
  if (FLAGS_gpu.size() == 0
      && solver_param.solver_mode() == caffe::SolverParameter_SolverMode_GPU) {
      if (solver_param.has_device_id()) {
          FLAGS_gpu = "" +
              boost::lexical_cast<string>(solver_param.device_id());
      } else {  // Set default GPU if unspecified
          FLAGS_gpu = "" + boost::lexical_cast<string>(0);
      }
  }

  vector<int> gpus;
  get_gpus(&gpus);
  if (gpus.size() == 0) {
    LOG(INFO) << "Use CPU.";
    Caffe::set_mode(Caffe::CPU);
  } else {
    ostringstream s;
    for (int i = 0; i < gpus.size(); ++i) {
      s << (i ? ", " : "") << gpus[i];
    }
    LOG(INFO) << "Using GPUs " << s.str();
#ifndef CPU_ONLY
    cudaDeviceProp device_prop;
    for (int i = 0; i < gpus.size(); ++i) {
      cudaGetDeviceProperties(&device_prop, gpus[i]);
      LOG(INFO) << "GPU " << gpus[i] << ": " << device_prop.name;
    }
#endif
    solver_param.set_device_id(gpus[0]);
    Caffe::SetDevice(gpus[0]);
    Caffe::set_mode(Caffe::GPU);
    Caffe::set_solver_count(gpus.size());
  }

  caffe::SignalHandler signal_handler(
        GetRequestedAction(FLAGS_sigint_effect),
        GetRequestedAction(FLAGS_sighup_effect));

  shared_ptr<caffe::Solver<float> >
      solver(caffe::SolverRegistry<float>::CreateSolver(solver_param));

  solver->SetActionFunction(signal_handler.GetActionFunction());

  if (FLAGS_snapshot.size()) {
    LOG(INFO) << "Resuming from " << FLAGS_snapshot;
    solver->Restore(FLAGS_snapshot.c_str());
  } else if (FLAGS_weights.size()) {
    CopyLayers(solver.get(), FLAGS_weights);
  }

  if (FLAGS_param_server != "") {
    LOG(INFO) << "Configuring multinode setup";

      if (FLAGS_param_server != "mpi") {
        LOG(ERROR) << "currently unsupported";
        return 1;
      }
      caffe::SynchronousNode<float> sync(solver, FLAGS_comm_threads);
      LOG(INFO) << "Starting Multi-node Optimization in mpi environment";
      sync.run();
  } else if (gpus.size() > 1) {
    caffe::P2PSync<float> sync(solver, NULL, solver->param());
    sync.Run(gpus);
  } else {
    LOG(INFO) << "Starting Optimization";
    solver->Solve();
  }
  LOG(INFO) << "Optimization Done.";
  return 0;
}
RegisterBrewFunction(train);

int data_server() {
  CHECK_GT(FLAGS_solver.size(), 0) << "Need a solver definition to train.";
  CHECK(!FLAGS_snapshot.size() || !FLAGS_weights.size())
      << "Give a snapshot to resume training or weights to finetune "
      "but not both.";

  caffe::SolverParameter solver_param;
  caffe::ReadSolverParamsFromTextFileOrDie(FLAGS_solver, &solver_param);

  caffe::SignalHandler signal_handler(
        GetRequestedAction(FLAGS_sigint_effect),
        GetRequestedAction(FLAGS_sighup_effect));

  shared_ptr<caffe::Solver<float> >
      solver(caffe::SolverRegistry<float>::CreateSolver(solver_param));

  solver->SetActionFunction(signal_handler.GetActionFunction());

  if (FLAGS_snapshot.size()) {
    LOG(INFO) << "Resuming from " << FLAGS_snapshot;
    solver->Restore(FLAGS_snapshot.c_str());
  } else if (FLAGS_weights.size()) {
    CopyLayers(solver.get(), FLAGS_weights);
  }
  LOG(INFO) << "Starting Data Server";
  caffe::DataServer<float> server(
    solver, FLAGS_listen_address, FLAGS_param_server, FLAGS_comm_threads);
  server.run();
  return 0;
}
RegisterBrewFunction(data_server);

// Test: score a model.
int test() {
  CHECK_GT(FLAGS_model.size(), 0) << "Need a model definition to score.";
  CHECK_GT(FLAGS_weights.size(), 0) << "Need model weights to score.";
  vector<string> stages = get_stages_from_flags();

  // Set device id and mode
  vector<int> gpus;
  get_gpus(&gpus);
  if (gpus.size() != 0) {
    LOG(INFO) << "Use GPU with device ID " << gpus[0];
#ifndef CPU_ONLY
    cudaDeviceProp device_prop;
    cudaGetDeviceProperties(&device_prop, gpus[0]);
    LOG(INFO) << "GPU device name: " << device_prop.name;
#endif
    Caffe::SetDevice(gpus[0]);
    Caffe::set_mode(Caffe::GPU);
  } else {
    LOG(INFO) << "Use CPU.";
    Caffe::set_mode(Caffe::CPU);
  }
  // Instantiate the caffe net.
  Net<float> caffe_net(FLAGS_model, caffe::TEST, FLAGS_level, &stages);
  caffe_net.CopyTrainedLayersFrom(FLAGS_weights);
  LOG(INFO) << "Running for " << FLAGS_iterations << " iterations.";

  vector<int> test_score_output_id;
  vector<float> test_score;
  float loss = 0;
  for (int i = 0; i < FLAGS_iterations; ++i) {
    float iter_loss;
    const vector<Blob<float>*>& result =
        caffe_net.Forward(&iter_loss);
    loss += iter_loss;
    int idx = 0;
    for (int j = 0; j < result.size(); ++j) {
      const float* result_vec = result[j]->cpu_data();
      for (int k = 0; k < result[j]->count(); ++k, ++idx) {
        const float score = result_vec[k];
        if (i == 0) {
          test_score.push_back(score);
          test_score_output_id.push_back(j);
        } else {
          test_score[idx] += score;
        }
        const std::string& output_name = caffe_net.blob_names()[
            caffe_net.output_blob_indices()[j]];
        LOG(INFO) << "Batch " << i << ", " << output_name << " = " << score;
      }
    }
  }
  loss /= FLAGS_iterations;
  LOG(INFO) << "Loss: " << loss;
  for (int i = 0; i < test_score.size(); ++i) {
    const std::string& output_name = caffe_net.blob_names()[
        caffe_net.output_blob_indices()[test_score_output_id[i]]];
    const float loss_weight = caffe_net.blob_loss_weights()[
        caffe_net.output_blob_indices()[test_score_output_id[i]]];
    std::ostringstream loss_msg_stream;
    const float mean_score = test_score[i] / FLAGS_iterations;
    if (loss_weight) {
      loss_msg_stream << " (* " << loss_weight
                      << " = " << loss_weight * mean_score << " loss)";
    }
    LOG(INFO) << output_name << " = " << mean_score << loss_msg_stream.str();
  }

  return 0;
}
RegisterBrewFunction(test);


// Time: benchmark the execution time of a model.
int time() {
  CHECK_GT(FLAGS_model.size(), 0) << "Need a model definition to time.";
  caffe::Phase phase = get_phase_from_flags(caffe::TRAIN);
  vector<string> stages = get_stages_from_flags();

  // Set device id and mode
  vector<int> gpus;
  get_gpus(&gpus);
  if (gpus.size() != 0) {
    LOG(INFO) << "Use GPU with device ID " << gpus[0];
    Caffe::SetDevice(gpus[0]);
    Caffe::set_mode(Caffe::GPU);
  } else {
    LOG(INFO) << "Use CPU.";
    Caffe::set_mode(Caffe::CPU);
  }
  // Instantiate the caffe net.
  Net<float> caffe_net(FLAGS_model, phase, FLAGS_level, &stages);

  // Do a clean forward and backward pass, so that memory allocation are done
  // and future iterations will be more stable.
  LOG(INFO) << "Performing Forward";
  // Note that for the speed benchmark, we will assume that the network does
  // not take any input blobs.
  float initial_loss;
  caffe_net.Forward(&initial_loss);
  LOG(INFO) << "Initial loss: " << initial_loss;
  LOG(INFO) << "Performing Backward";
  caffe_net.Backward();

  const vector<shared_ptr<Layer<float> > >& layers = caffe_net.layers();
  const vector<vector<Blob<float>*> >& bottom_vecs = caffe_net.bottom_vecs();
  const vector<vector<Blob<float>*> >& top_vecs = caffe_net.top_vecs();
  const vector<vector<bool> >& bottom_need_backward =
      caffe_net.bottom_need_backward();
  LOG(INFO) << "*** Benchmark begins ***";
  LOG(INFO) << "Testing for " << FLAGS_iterations << " iterations.";
  Timer total_timer;
  total_timer.Start();
  Timer forward_timer;
  Timer backward_timer;
  Timer timer;
  std::vector<double> forward_time_per_layer(layers.size(), 0.0);
  std::vector<double> backward_time_per_layer(layers.size(), 0.0);
  double forward_time = 0.0;
  double backward_time = 0.0;
  for (int j = 0; j < FLAGS_iterations; ++j) {
    Timer iter_timer;
    iter_timer.Start();
    forward_timer.Start();
    for (int i = 0; i < layers.size(); ++i) {
      timer.Start();
      layers[i]->Forward(bottom_vecs[i], top_vecs[i]);
      forward_time_per_layer[i] += timer.MicroSeconds();
    }
    forward_time += forward_timer.MicroSeconds();
    backward_timer.Start();
    for (int i = layers.size() - 1; i >= 0; --i) {
      timer.Start();
      layers[i]->Backward(top_vecs[i], bottom_need_backward[i],
                          bottom_vecs[i]);
      backward_time_per_layer[i] += timer.MicroSeconds();
    }
    backward_time += backward_timer.MicroSeconds();
    LOG(INFO) << "Iteration: " << j + 1 << " forward-backward time: "
      << iter_timer.MilliSeconds() << " ms.";
  }
  LOG(INFO) << "Average time per layer: ";
  for (int i = 0; i < layers.size(); ++i) {
    const caffe::string& layername = layers[i]->layer_param().name();
    LOG(INFO) << std::setfill(' ') << std::setw(10) << layername <<
      "\tforward: " << forward_time_per_layer[i] / 1000 /
      FLAGS_iterations << " ms.";
    LOG(INFO) << std::setfill(' ') << std::setw(10) << layername  <<
      "\tbackward: " << backward_time_per_layer[i] / 1000 /
      FLAGS_iterations << " ms.";
  }
  total_timer.Stop();
  LOG(INFO) << "Average Forward pass: " << forward_time / 1000 /
    FLAGS_iterations << " ms.";
  LOG(INFO) << "Average Backward pass: " << backward_time / 1000 /
    FLAGS_iterations << " ms.";
  LOG(INFO) << "Average Forward-Backward: " << total_timer.MilliSeconds() /
    FLAGS_iterations << " ms.";
  LOG(INFO) << "Total Time: " << total_timer.MilliSeconds() << " ms.";
  LOG(INFO) << "*** Benchmark ends ***";
  return 0;
}
RegisterBrewFunction(time);


// collect & compare: Debugging extansion for CPU-GPU functional comparison
#include <stdio.h>
typedef float real_t;

void getFileName(char *file_name, bool use_gpu, const char *name, int id) {
  const char *prefix = use_gpu ? "GPU" : "CPU";
  snprintf(file_name, FILENAME_MAX, "%s%s%04i.bin", prefix, name, id);
}

bool saveToFile(bool use_gpu, const char *name, int id,
    const real_t *data, unsigned count) {
  char file_name[FILENAME_MAX];
  getFileName(file_name, use_gpu, name, id);

  FILE *file = fopen(file_name, "w+b");
  if (!file) {
    LOG(ERROR) << "Failed to create file '" << file_name << "'.";
    return false;
  }

  size_t bytesToWrite = count * sizeof(data[0]);
  size_t bytesWritten = fwrite(data, 1, bytesToWrite, file);
  fclose(file);

  if (bytesWritten != bytesToWrite) {
    LOG(ERROR) << "Failed to write data to '" << file_name << "' file.";
    return false;
  }

  return true;
}

bool loadFromFile(bool use_gpu, const char *name, int id,
    real_t *data, unsigned count) {
  char file_name[FILENAME_MAX];
  getFileName(file_name, use_gpu, name, id);

  FILE *file = fopen(file_name, "rb");
  if (!file) {
    LOG(ERROR) << "Failed to open file '" << file_name << "' for read.";
    return false;
  }

  size_t bytesToRead = count * sizeof(data[0]);
  size_t bytesRead = fread(data, 1, bytesToRead, file);
  fclose(file);

  if (bytesRead != bytesToRead) {
    LOG(ERROR) << "Failed to read data from '" << file_name << "' file.";
    return false;
  }

  return true;
}

int collect() {
  #ifndef DETERMINISTIC
    LOG(ERROR) << "Recompile caffe with DETERMINISTIC to run collect tool";
    return 1;
  #endif
  CHECK_GT(FLAGS_model.size(), 0) << "Need a model definition!";

  vector<int> gpus;
  get_gpus(&gpus);
  bool use_gpu = (gpus.size() != 0);
  if (use_gpu) {
    LOG(INFO) << "Use GPU with device ID " << gpus[0];
    Caffe::SetDevice(gpus[0]);
    Caffe::set_mode(Caffe::GPU);
  } else {
    LOG(INFO) << "Use CPU.";
    Caffe::set_mode(Caffe::CPU);
  }

  Net<real_t> caffe_net(FLAGS_model, caffe::TRAIN);
  const vector<shared_ptr<Layer<real_t> > >& layers = caffe_net.layers();
  const vector<shared_ptr<Blob<real_t> > >& params = caffe_net.params();
  const vector<vector<Blob<real_t>*> >& bottom_vecs = caffe_net.bottom_vecs();
  const vector<vector<Blob<real_t>*> >& top_vecs = caffe_net.top_vecs();
  const vector<vector<bool> >& bottom_need_backward =
    caffe_net.bottom_need_backward();

  FILE *infoFile = fopen(use_gpu ? "GPUInfo.txt" : "CPUInfo.txt", "w+t");
  LOG(INFO) << "*** Collect procedure begins ***";

  for (int i = 0; i < params.size(); i++) {
    caffe::caffe_set(params[i]->count(), 0.f,
      params[i]->mutable_cpu_diff());
  }

  for (int i = 0; i < layers.size(); ++i) {
    LOG(INFO) << "Collecting FW Layer[" << i << "]: " << layers[i]->type();
    fprintf(infoFile, "Fwrd%04i: %s\n", i, layers[i]->type());
    layers[i]->Forward(bottom_vecs[i], top_vecs[i]);
    saveToFile(use_gpu, "Fwrd", i,
      top_vecs[i][0]->cpu_data(), top_vecs[i][0]->count());
  }

  for (int i = layers.size() - 1; i >= 0; --i) {
    LOG(INFO) << "Collecting BW Layer[" << i << "]: " << layers[i]->type();
    fprintf(infoFile, "Bwrd%04i: %s\n", i, layers[i]->type());
    layers[i]->Backward(top_vecs[i], bottom_need_backward[i], bottom_vecs[i]);
    if (bottom_need_backward[i][0]) {
      saveToFile(use_gpu, "Bwrd", i,
        bottom_vecs[i][0]->cpu_diff(), bottom_vecs[i][0]->count());
    }
  }

  LOG(INFO) << "Collecting gradients and weights";
  for (int i = 0; i < params.size(); i++) {
    saveToFile(use_gpu, "Grad", i,
      params[i]->cpu_diff(), params[i]->count());
    saveToFile(use_gpu, "Wght", i,
      params[i]->cpu_data(), params[i]->count());
  }

  LOG(INFO) << "*** Collect procedure ends ***";
  fclose(infoFile);
  return 0;
}
RegisterBrewFunction(collect);

int compare() {
  #ifndef DETERMINISTIC
    LOG(ERROR) << "Recompile caffe with DETERMINISTIC to run compare tool";
    return 1;
  #endif
  CHECK_GT(FLAGS_model.size(), 0) << "Need a model definition!";

  vector<int> gpus;
  get_gpus(&gpus);
  bool use_gpu = (gpus.size() != 0);
  if (use_gpu) {
    LOG(INFO) << "Use GPU with device ID " << gpus[0];
    Caffe::SetDevice(gpus[0]);
    Caffe::set_mode(Caffe::GPU);
  } else {
    LOG(INFO) << "Use CPU.";
    Caffe::set_mode(Caffe::CPU);
  }

  Net<real_t> caffe_net(FLAGS_model, caffe::TRAIN);
  const vector<shared_ptr<Layer<real_t> > >& layers = caffe_net.layers();
  const vector<shared_ptr<Blob<real_t> > >& params = caffe_net.params();
  const vector<vector<Blob<real_t>*> >& bottom_vecs = caffe_net.bottom_vecs();
  const vector<vector<Blob<real_t>*> >& top_vecs = caffe_net.top_vecs();
  const vector<vector<bool> >& bottom_need_backward =
    caffe_net.bottom_need_backward();

  FILE *infoFile = fopen(use_gpu ? "GPUInfo.txt" : "CPUInfo.txt", "w+t");
  LOG(INFO) << "*** Compare procedure begins ***";

  for (int i = 0; i < params.size(); i++) {
    caffe::caffe_set(params[i]->count(), 0.f,
      params[i]->mutable_cpu_diff());
  }

  for (int i = 0; i < layers.size(); ++i) {
    LOG(INFO) << "Collecting FW Layer[" << i << "]: " << layers[i]->type();
    fprintf(infoFile, "Fwrd%04i: %s\n", i, layers[i]->type());
    layers[i]->Forward(bottom_vecs[i], top_vecs[i]);
    saveToFile(use_gpu, "Fwrd", i,
      top_vecs[i][0]->cpu_data(), top_vecs[i][0]->count());
    loadFromFile(!use_gpu, "Fwrd", i,
      top_vecs[i][0]->mutable_cpu_data(), top_vecs[i][0]->count());
  }

  for (int i = layers.size() - 1; i >= 0; --i) {
    LOG(INFO) << "Collecting BW Layer[" << i << "]: " << layers[i]->type();
    fprintf(infoFile, "Bwrd%04i: %s\n", i, layers[i]->type());
    layers[i]->Backward(top_vecs[i], bottom_need_backward[i], bottom_vecs[i]);
    if (bottom_need_backward[i][0]) {
      saveToFile(use_gpu, "Bwrd", i,
        bottom_vecs[i][0]->cpu_diff(), bottom_vecs[i][0]->count());
      loadFromFile(!use_gpu, "Bwrd", i,
        bottom_vecs[i][0]->mutable_cpu_diff(), bottom_vecs[i][0]->count());
    }
  }

  LOG(INFO) << "Collecting gradients and weights";
  for (int i = 0; i < params.size(); i++) {
    saveToFile(use_gpu, "Grad", i,
      params[i]->cpu_diff(), params[i]->count());
    saveToFile(use_gpu, "Wght", i,
      params[i]->cpu_data(), params[i]->count());
  }

  LOG(INFO) << "*** Compare procedure ends ***";
  fclose(infoFile);
  return 0;
}
RegisterBrewFunction(compare);


int main(int argc, char** argv) {
  caffe::internode::mpi_init(argc, argv);
  // Print output to stderr (while still logging).
  FLAGS_alsologtostderr = 1;
  // Set version
  gflags::SetVersionString(AS_STRING(CAFFE_VERSION));
  // Usage message.
  gflags::SetUsageMessage("command line brew\n"
      "usage: caffe <command> <args>\n\n"
      "commands:\n"
      "  train           train or finetune a model\n"
      "  test            score a model\n"
      "  data_server     run data server - remote data source\n"
      "  device_query    show GPU diagnostic information\n"
      "  time            benchmark model execution time\n"
      "  collect         collects layer data on specified device\n"
      "  compare         collects layer data using inputs from other device");
  // Run tool or show usage.
  caffe::GlobalInit(&argc, &argv);
  if (argc == 2) {
#ifdef WITH_PYTHON_LAYER
    try {
#endif
      int ret = GetBrewFunction(caffe::string(argv[1]))();
      caffe::internode::mpi_finalize();
      return ret;
#ifdef WITH_PYTHON_LAYER
    } catch (bp::error_already_set) {
      PyErr_Print();
      caffe::internode::mpi_finalize();
      return 1;
    }
#endif
  } else {
    gflags::ShowUsageWithFlagsRestrict(argv[0], "tools/caffe");
  }
  caffe::internode::mpi_finalize();
  return 0;
}
